File: frontend/lib/omr.ts
import { apiFetch, apiFetchBlob } from '@/lib/api';

export type OMRScanStatus = 'processing' | 'success' | 'review_needed' | 'failed';

export interface OMRTemplate {
  id: string;
  exam_id: string | null;
  layout_version: string;
  total_questions: number;
  options_per_question: number;
  correct_answers: Record<string, string> | null;
  created_at: string;
  updated_at: string;
}

export interface OMRScan {
  id: string;
  omr_template_id: string;
  student_code: string | null;
  student_id: string | null;
  status: OMRScanStatus;
  image_url: string;
  detected_answers: Record<string, string | null> | null;
  raw_confidence: Record<string, unknown> | null;
  score: string | number | null;
  error_message: string | null;
  processed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface Grade {
  id: string;
  student_id: string;
  source_type: string;
  source_id: string;
  score: string | number;
  teacher_id: string;
  created_at: string;
  updated_at: string;
}

export interface CreateTemplateInput {
  layout_version: string;
  total_questions: number;
  options_per_question?: number;
  correct_answers: Record<string, string>;
  exam_id?: string | null;
}

export function listTemplates() {
  return apiFetch<OMRTemplate[]>('/omr/templates');
}

export function getTemplate(templateId: string) {
  return apiFetch<OMRTemplate>(`/omr/templates/${templateId}`);
}

export function createTemplate(input: CreateTemplateInput) {
  return apiFetch<OMRTemplate>('/omr/templates', {
    method: 'POST',
    body: JSON.stringify(input),
  });
}

export function downloadTemplatePdf(templateId: string, studentCode: string) {
  const query = studentCode ? `?student_code=${encodeURIComponent(studentCode)}` : '';
  return apiFetchBlob(`/omr/templates/${templateId}/pdf${query}`);
}

export function downloadTemplatePreview(templateId: string, studentCode: string) {
  const query = studentCode ? `?student_code=${encodeURIComponent(studentCode)}` : '';
  return apiFetchBlob(`/omr/templates/${templateId}/preview.png${query}`);
}

export async function uploadScan(templateId: string, file: File) {
  const form = new FormData();
  form.append('omr_template_id', templateId);
  form.append('file', file);
  return apiFetch<OMRScan>('/omr/scans/upload', {
    method: 'POST',
    body: form,
  });
}

export function getScan(scanId: string) {
  return apiFetch<OMRScan>(`/omr/scans/${scanId}`);
}

export function updateScan(
  scanId: string,
  payload: { student_code?: string; detected_answers?: Record<string, string | null> },
) {
  return apiFetch<OMRScan>(`/omr/scans/${scanId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  });
}

export function confirmScan(scanId: string) {
  return apiFetch<Grade>(`/omr/scans/${scanId}/confirm`, {
    method: 'POST',
  });
}

File: backend/app/schemas/omr.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.enums import GradeSourceType, OMRScanStatus

# --- OMR Template Schemas ---


class OMRTemplateBase(BaseModel):
    exam_id: Optional[UUID] = None
    layout_version: str
    total_questions: int
    options_per_question: int = 5


class OMRTemplateCreate(OMRTemplateBase):
    correct_answers: Optional[Dict[str, str]] = None


class OMRTemplateResponse(OMRTemplateBase):
    id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- OMR Scan Schemas ---


class OMRScanBase(BaseModel):
    omr_template_id: UUID
    student_code: Optional[str] = None
    student_id: Optional[UUID] = None
    status: OMRScanStatus
    image_url: str
    detected_answers: Optional[Dict[str, Optional[str]]] = None
    raw_confidence: Optional[Dict[str, Any]] = None
    score: Optional[Decimal] = None
    error_message: Optional[str] = None
    processed_at: Optional[datetime] = None


class OMRScanResponse(OMRScanBase):
    id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class OMRScanUpdate(BaseModel):
    student_code: Optional[str] = None
    detected_answers: Optional[Dict[str, Optional[str]]] = None
    status: Optional[OMRScanStatus] = None


# --- Grade Schemas ---


class GradeBase(BaseModel):
    student_id: UUID
    source_type: GradeSourceType
    source_id: UUID
    score: Decimal
    teacher_id: UUID


class GradeCreate(GradeBase):
    pass


class GradeResponse(GradeBase):
    id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
